sklearn/utils/validation.py

```diff
@@ -465,10 +465,10 @@ def check_consistent_length(*arrays):
     >>> b = [2, 3, 4]
     >>> check_consistent_length(a, b)
     """
-
-    lengths = [_num_samples(X) for X in arrays if X is not None]
-    uniques = np.unique(lengths)
-    if len(uniques) > 1:
+    xp, _ = get_namespace(*arrays)
+    lengths = xp.asarray([_num_samples(X) for X in arrays if X is not None])
+    uniques = xp.unique_values(lengths)
+    if uniques.shape[0] > 1:
```

Comment:
While it's possible to change the code of this function, I don't expect any computational or operational benefit in doing so. The original code should already accept mixed array API compatible arrays and other non-array API compatible data-structures because `_num_samples` is already array API compatible.

The number of arrays passed to `check_consistent_length` is typically 2 or 3. So using numpy is perfectly fine. We could even directly use:

```python
if len(set(lengths)) > 1:
   ...
```

to not even rely on numpy at all in this code.

I would therefore revert to the original code of this function (without the `get_namespace` call), or switch to use a simple set and make this PR only add the new test that checks that `check_consistent_lengths` works as expected with array API compatible inputs when dispatch is enabled.

Reply:
Not calling `get_namespace(*arrays)` would also bring the short term benefit of decoupling this PR from the review of https://github.com/scikit-learn/scikit-learn/pull/29476.